src/sagemaker/workflow/pipeline.py

```diff
@@ -322,6 +322,16 @@ def _interpolate_step_collection_name_in_depends_on(self, step_requests: dict):
                     depends_on.append(depend_step_name)
             step_request["DependsOn"] = depends_on
 
+    def _generate_adjacency_list(self):
+        """Generate an adjacency list representing the step dependency DAG in this pipeline."""
+        adjacency_list = {}
+        for step in self.steps:
+            if isinstance(step, Step):
+                adjacency_list[step.name] = step._find_step_dependencies()
+            elif isinstance(step, StepCollection):
+                adjacency_list.update(step._find_step_dependencies())
+        return adjacency_list
```

Comment:
Suggest we create a new class and offload the building of the graph, validation (no cycle), traversal to this class

```
class _PipelineGraph:
    def __init__ (self, steps: List[Step]): 
        self.adjacency_list = {}
        ...
    def __iter__():
        ...
    def __next()__() -> List[Step]:
        ...

```

Reply:
Condition step needs to be treated in a special way. See
https://code.amazon.com/packages/IronmanTiogaPipelineDefinitionRepository/blobs/8b985bed9418f5cace3f7bea553e4354752c61f3/--/src/com/amazon/ironmantiogapipelinerepository/definition/PipelineDefinition.java#L73